File: tests/fixtures/go/data_isolation.go
package main

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
)

// --- D1: DB write detection ---

func CreateProject(ctx context.Context, pool *pgxpool.Pool, name string, userID string) error {
	// Should be detected as DbWriteRef (INSERT INTO projects)
	_, err := pool.Exec(ctx, "INSERT INTO projects (name, user_id) VALUES ($1, $2)", name, userID)
	return err
}

func UpdateTask(ctx context.Context, pool *pgxpool.Pool, taskID string, status string) error {
	// Should be detected as DbWriteRef (UPDATE tasks)
	_, err := pool.Exec(ctx, "UPDATE tasks SET status = $1 WHERE id = $2", status, taskID)
	return err
}

func DeleteAsset(ctx context.Context, pool *pgxpool.Pool, assetID string) error {
	// Should be detected as DbWriteRef (DELETE FROM assets)
	_, err := pool.Exec(ctx, "DELETE FROM assets WHERE id = $1", assetID)
	return err
}

// --- D2: RLS activation ---

func WithRLS(ctx context.Context, pool *pgxpool.Pool, userID string) error {
	// Should be detected as RlsContextRef
	_, err := pool.Exec(ctx, fmt.Sprintf("SET LOCAL app.current_user_id = '%s'", userID))
	return err
}

// --- D4: Missing ownership filter ---

func GetTaskByID(ctx context.Context, pool *pgxpool.Pool, taskID string) error {
	// SELECT without user_id filter — D5 should flag
	row := pool.QueryRow(ctx, "SELECT * FROM tasks WHERE id = $1", taskID)
	_ = row
	return nil
}

// --- D6: Redis cache-only ---

func CacheSessionState(ctx context.Context, rdb *redis.Client, sessionID string, data string) {
	// Redis SET with TTL, no DB write — D6 should flag
	rdb.Set(ctx, fmt.Sprintf("session:state:%s", sessionID), data, 86400)
}

func GetSessionState(ctx context.Context, rdb *redis.Client, sessionID string) string {
	// Redis GET — should be detected as RedisOp::Read
	val, _ := rdb.Get(ctx, fmt.Sprintf("session:state:%s", sessionID)).Result()
	return val
}

// --- D7: Hardcoded credentials ---

type Config struct {
	MinioAccessKey   string
	MinioSecretKey   string
	FederationSecret string
}

var defaultConfig = Config{
	MinioAccessKey:   "minioadmin",
	MinioSecretKey:   "minioadmin",
	FederationSecret: "federation_secret",
}

const InternalAPIKey = "dev-internal-key"

File: tests/fixtures/typescript/data_isolation.ts
/**
 * Fixture for S12 Data Isolation Audit — TypeScript patterns.
 */

import { Pool } from 'pg';
import Redis from 'ioredis';

const pool = new Pool();
const redis = new Redis();

// --- D1: DB write detection ---

export async function createUser(name: string, email: string) {
  // Should be detected as DbWriteRef (INSERT INTO users)
  await pool.query(
    `INSERT INTO users (name, email) VALUES ($1, $2)`,
    [name, email]
  );
}

export async function updateOrder(orderId: string, status: string) {
  // Should be detected as DbWriteRef (UPDATE orders)
  await pool.query(
    `UPDATE orders SET status = $1 WHERE id = $2`,
    [status, orderId]
  );
}

export async function deleteSession(sessionId: string) {
  // Should be detected as DbWriteRef (DELETE FROM sessions)
  await pool.query(`DELETE FROM sessions WHERE id = $1`, [sessionId]);
}

// --- D4: Missing ownership filter ---

export async function getOrderById(orderId: string) {
  // SELECT without user_id filter — D5 should flag this
  const result = await pool.query(
    `SELECT * FROM orders WHERE id = $1`,
    [orderId]
  );
  return result.rows[0];
}

export async function updateOrderUnsafe(orderId: string, data: any) {
  // UPDATE without user_id filter — D4 should flag this
  await pool.query(
    `UPDATE orders SET data = $1 WHERE id = $2`,
    [data, orderId]
  );
}

// --- D6: Redis cache-only ---

export async function cacheUserProfile(userId: string, profile: object) {
  // Redis SET with TTL, no DB write nearby — D6 should flag
  await redis.set(`user:profile:${userId}`, JSON.stringify(profile), 'EX', 3600);
}

export async function getUserProfile(userId: string) {
  // Redis GET — should be detected as RedisOp::Read
  return redis.get(`user:profile:${userId}`);
}

// --- D7: Hardcoded credentials ---

const config = {
  database_password: "postgres123",
  redis_secret: "redis-dev-secret",
  jwt_token: process.env.JWT_SECRET || "",  // env ref, should NOT be flagged
};

export const INTERNAL_API_KEY = "dev-internal-key-12345";

// --- D8: RLS context setting ---

export async function setRlsContext(userId: string) {
  // SET LOCAL in a template literal — should be detected as RlsContextRef
  await pool.query(`SET LOCAL app.current_user_id = '${userId}'`);
}

export async function setRlsViaConfig(tenantId: string) {
  // set_config call — should also be detected
  await pool.query(`SELECT set_config('app.current_tenant_id', $1, true)`, [tenantId]);
}
